### PR TITLE
[CK_TILE] Add operator batched_transpose

### PR DESCRIPTION
**Pull request by fangche123**

### CODE REVIEW

File: example/ck_tile/35_batched_transpose/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,3 @@
+add_executable(tile_example_batched_transpose EXCLUDE_FROM_ALL batched_transpose.cpp batched_transpose_api.cpp)
+target_include_directories(tile_example_batched_transpose PRIVATE ${CMAKE_CURRENT_SOURCE_DIR}/)
+target_compile_options(tile_example_batched_transpose PRIVATE -v --save-temps -Wno-gnu-line-marker)
```
Comment (poyenc) [inline example/ck_tile/35_batched_transpose/CMakeLists.txt]: Are those options essential? Or they are only used for debuging?

**Comment by poyenc** [comment]
I just pushed re-formated files. Develops have to run remod.py script under **include/ck_tile** & **example/ck_tile** folders. 
The script will re-format source files and generate interface headers which include sub-directory headers.

File: example/ck_tile/35_batched_transpose/batched_transpose_api.cpp [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+                        ck_tile::stream_config s)
+{
+    if(t.type == "fp16")
+    {
+        using ts_type = ck_tile::fp16_t;
+        if(t.layout == "NCHW")
+        {
+            BATCHED_TRANSPOSE_DISPATCH()
+        }
+        else if(t.layout == "NHWC")
```
Comment (poyenc) [inline example/ck_tile/35_batched_transpose/batched_transpose_api.cpp]: Are those checks necessary? Either way you end-up calling the `BATCHED_TRANSPOSE_DISPATC()` macro

File: example/ck_tile/35_batched_transpose/batched_transpose_api.hpp [hunk 3]
```diff
@@ -0,0 +1,22 @@
+// SPDX-License-Identifier: MIT
+// Copyright (c) 2018-2023, Advanced Micro Devices, Inc. All rights reserved.
```
Comment (poyenc) [inline example/ck_tile/35_batched_transpose/batched_transpose_api.hpp]: update year.

File: example/ck_tile/35_batched_transpose/batched_transpose_api.cpp [hunk 4]
```diff
@@ -0,0 +1,76 @@
+#include "batched_transpose_api.hpp"
```
Comment (poyenc) [inline example/ck_tile/35_batched_transpose/batched_transpose_api.cpp]: missed the copyright

File: example/ck_tile/35_batched_transpose/CMakeLists.txt (hunk 1 above)
Comment (poyenc) [inline example/ck_tile/35_batched_transpose/CMakeLists.txt]: you may need to define `TEST_BATCHED_TRANSPOSE_VERBOSE` macro here for testing.

File: example/ck_tile/35_batched_transpose/batched_transpose_api.hpp (hunk 3 above)
Comment (fangche123) [reply example/ck_tile/35_batched_transpose/batched_transpose_api.hpp]: ok, I will fix it.

File: include/ck_tile/host/reference/reference_batched_transpose.hpp [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+#include <thread>
+
+namespace ck_tile {
+
+template <typename Type>
+CK_TILE_HOST void reference_batched_transpose(const HostTensor<Type>& x,
+                                              HostTensor<Type>& y,
+                                              std::string layout = "NCHW")
+{
+    // index_t rank = x.get_num_of_dimension();
```
Comment (poyenc) [inline include/ck_tile/host/reference/reference_batched_transpose.hpp]: Accidentally commented out?

File: example/ck_tile/35_batched_transpose/CMakeLists.txt (hunk 1 above)
Comment (fangche123) [reply example/ck_tile/35_batched_transpose/CMakeLists.txt]: Yes, these options are temporarily used for debug

File: include/ck_tile/host/reference/reference_batched_transpose.hpp [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+#pragma once
+
+#include "ck_tile/core.hpp"
+#include "ck_tile/host/host_tensor.hpp"
+#include <thread>
+
+namespace ck_tile {
+
+template <typename Type>
+CK_TILE_HOST void reference_batched_transpose(const HostTensor<Type>& x,
```
Comment (poyenc) [inline include/ck_tile/host/reference/reference_batched_transpose.hpp:13]: please document for the output layout.

File: include/ck_tile/host/reference/reference_batched_transpose.hpp [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+                                              HostTensor<Type>& y,
```
Comment (poyenc) [inline include/ck_tile/host/reference/reference_batched_transpose.hpp]: Please document for the output layout.

File: include/ck_tile/ops/batched_transpose/kernel/batched_transpose_kernel.hpp [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+        k.dim_h      = h.dim_h;
+        k.dim_w      = h.dim_w;
+        return k;
+    }
+
+    CK_TILE_HOST_DEVICE static constexpr auto BlockSize() { return Problem::kBlockSize; }
+
+    CK_TILE_DEVICE void operator()(Kargs kargs) const
+    {
+        printf("in kernel. k.height:%d k.width:%d\n", kargs.height, kargs.width);
```
Comment (poyenc) [inline include/ck_tile/ops/batched_transpose/kernel/batched_transpose_kernel.hpp]: Forgot to remove debug code?

File: example/ck_tile/35_batched_transpose/batched_transpose.cpp [hunk 9]
```diff
@@ -0,0 +1,310 @@
...
+        unsigned max_rounding_point_distance = 1;
+        double atol                          = 0.0625;
+        return ck_tile::make_tuple(max_rounding_point_distance, atol);
+    }
+}
+
+auto create_args(int argc, char* argv[])
+{
+    ck_tile::ArgParser arg_parser;
+    arg_parser.insert("v", "1", "weather do CPU validation or not")
```
Comment (ThomasNing) [inline example/ck_tile/35_batched_transpose/batched_transpose.cpp]: typo: whether

File: example/ck_tile/35_batched_transpose/batched_transpose.cpp [hunk 10]
```diff
@@ -0,0 +1,310 @@
...
+#include "ck_tile/ops/reduce.hpp"
+#include "batched_transpose_api.hpp"
+#include "batched_transpose.hpp"
+
+#ifndef TEST_BATCHED_TRANSPOSE_VERBOSE
+#define TEST_BATCHED_TRANSPOSE_VERBOSE 1
+#endif
+
+template <typename T>
+void dump_host_tensor_4d(const ck_tile::HostTensor<T>& x)
```
Comment (ThomasNing) [inline example/ck_tile/35_batched_transpose/batched_transpose.cpp]: Please comment this out or put that to another file as a helper debug function. We could get a debug option in the argument input. WDYT?

File: example/ck_tile/35_batched_transpose/batched_transpose.cpp [hunk 11]
```diff
@@ -0,0 +1,310 @@
...
+    int C            = args.get_int("C");
+    int H            = args.get_int("H");
+    int W            = args.get_int("W");
+    int stride_dim0  = args.get_int("stride_dim0");
+    int stride_dim1  = args.get_int("stride_dim1");
+    int stride_dim2  = args.get_int("stride_dim2");
+    int stride_dim3  = args.get_int("stride_dim3");
+    int seed         = args.get_int("seed");
+
+    std::string layout_in = "NCHW", layout_out = "NHWC";
```
Comment (ThomasNing) [inline example/ck_tile/35_batched_transpose/batched_transpose.cpp]: Do we only support one kind of the memory layout from "NCHW" to "NHWC"? Theoretically, we could also support NWCH with another round of the transpose.

File: example/ck_tile/35_batched_transpose/batched_transpose.cpp [hunk 12]
```diff
@@ -0,0 +1,310 @@
...
+    int stride_out_dim0 = stride_dim0, stride_out_dim1 = W * C, stride_out_dim2 = C,
+        stride_out_dim3 = 1;
+    if(stride_dim0 == C * H * W && stride_dim1 == 1 && stride_dim2 == W * C && stride_dim3 == C)
```
Comment (ThomasNing) [inline example/ck_tile/35_batched_transpose/batched_transpose.cpp]: We should decide the stride configuration depending on the layout configuration right?

File: example/ck_tile/35_batched_transpose/batched_transpose.hpp [hunk 13]
```diff
@@ -0,0 +1,130 @@
+// SPDX-License-Identifier: MIT
+// Copyright (c) 2018-2024, Advanced Micro Devices, Inc. All rights reserved.
+
+#include <vector>
+#include <string>
+
+#pragma once
+
+struct transpose_kernel_param_t
```
Comment (ThomasNing) [inline example/ck_tile/35_batched_transpose/batched_transpose.hpp]: Comments on the pack_x pack_y & ediv_x ediv_y?

File: example/ck_tile/35_batched_transpose/batched_transpose_api.cpp [hunk 14]
```diff
@@ -0,0 +1,76 @@
...
+    float ave_time = ck_tile::launch_kernel(                                           \
+        s, ck_tile::make_kernel<blocks.x, 1>(kernel{}, grids, blocks, 0, kargs));      \
+                                                                                       \
+    return ave_time;
+
+float batched_transpose(batched_transpose_trait t,
+                        batched_transpose_kargs a,
+                        ck_tile::stream_config s)
+{
+    if(t.type == "fp16")
```
Comment (ThomasNing) [inline example/ck_tile/35_batched_transpose/batched_transpose_api.cpp]: You could put the type as a template in the transpose_calc function. PTAL gemm_kernel as a reference.

**Pull request closed without merging**